File: 905-sort-array-by-parity/905-sort-array-by-parity.java
class Solution {
    public int[] sortArrayByParity(int[] nums) {
        int[] ans = new int[nums.length];
        int lo = 0, hi = nums.length - 1;
        
        for (int i = 0; i < nums.length; i++) 
            if ((nums[i] & 1) == 1) 
                ans[hi--] = nums[i];
            else
                ans[lo++] = nums[i];
        
        return ans;
    }
}

File: 905-sort-array-by-parity/905-sort-array-by-parity.cpp
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        for (int N = nums.size(), i = 0, j;; swap(nums[i], nums[j])) {
            for (; i < N && !(nums[i] & 1); ++i);
            for (j = i + 1; j < N && nums[j] & 1; ++j);
            if (j >= N) 
                break;
        }
        return nums;
    }
};